interp/cubic.go

```diff
@@ -297,3 +297,68 @@ func fritschButlandEdgeDerivative(xs, ys, slopes []float64, leftEdge bool) float
 	}
 	return g
 }
+
+// CubicSpline manage necessary equation for Natural cubic spline.
+type CubicSpline struct {
+	xs []float64
+
+	coeff *mat.Dense
+}
+
+// Fit fits a predictor to (X, Y) value pairs provided as two slices.
+func (cs *CubicSpline) Fit(xs, xy []float64) error {
+	n := len(xs)
+	cs.coeff = mat.NewDense((n-1)*4, 1, nil)
+	coeffs := mat.NewDense((n-1)*4, (n-1)*4, nil)
+	coeffsinv := mat.NewDense((n-1)*4, (n-1)*4, nil)
+	sy := mat.NewDense((n-1)*4, 1, nil)
+	cs.xs = xs
+	for i := 0; i < n-1; i++ {
+		for j := 0; j < 4; j++ {
+			coeffs.Set(i, 4*i+j, math.Pow(xs[i], float64(j)))
+			coeffs.Set(i+n-1, 4*i+j, math.Pow(xs[i+1], float64(j)))
+			if i < n-2 {
+				coeffs.Set(i+2*(n-1), 4*i+j, float64(j)*math.Pow(xs[i+1], float64(j-1)))
+				coeffs.Set(i+2*(n-1), 4*(i+1)+j, float64(-j)*math.Pow(xs[i+1], float64(j-1)))
+			}
+		}
+		if i < n-2 {
+			coeffs.Set(i+(n-1)*3-1, 4*i+3, float64(3)*xs[i+1])
+			coeffs.Set(i+(n-1)*3-1, 4*i+2, float64(1))
+			coeffs.Set(i+(n-1)*3-1, 4*(i+1)+3, float64(-3)*xs[i+1])
+			coeffs.Set(i+(n-1)*3-1, 4*(i+1)+2, float64(-1))
+		}
+		sy.Set(i, 0, xy[i])
+		sy.Set(i+n-1, 0, xy[i+1])
+	}
+	coeffs.Set(4*(n-1)-2, 3, float64(3)*xs[0])
+	coeffs.Set(4*(n-1)-2, 2, float64(1))
+	coeffs.Set(4*(n-1)-1, 4*(n-1)-1, float64(3)*xs[n-1])
+	coeffs.Set(4*(n-1)-1, 4*(n-1)-2, float64(1))
+	coeffsinv.Inverse(coeffs)
```

Comment:
This should really be a solve, rather than a multiply with an inverse, with an appropriate change of the type's field to being a `mat.VecDense` (concrete, not pointer), this can become `return cs.coeff.SolveVec(coeffs, &cs.coeff)`, with the opening preamble becoming `cs.coeff.ReuseAsVec((n - 1) * 4)` rather than the allocation that's there (and using it for both the temporary `sy` that's here and the final result).

Also note that there are `float64` conversions of untyped constants that are unnecessary in a few places above.